### PR TITLE
First back-end

### PR DESCRIPTION
**Pull request by kolayne**
Hi! This is the first version of the project's back-end. I've been trying to reduce its size as far as possible, however, it turned out to be somewhat large, so probably it's a good idea not to review the whole PR in one sitting and review it by parts several times instead, but it's totally up to you.

So, what is in the PR?
- Updates of some meta files (added usage to README.md, created .gitignore and a config sample file)
- Some basic HTML templates (which are most likely going to be rewritten completely, but I've added them to make it easier to test/debug the app. They may also be useful for people, who are going to create the UI: they may want to see sample templates to understand, how to create theirs)
- SQL bridge ~~(in fact, most of the file was copied from another solution I've found on the internet. Note that it uses the `reflect` package of go, which might be hard to understand without reading some docs, so if you hadn't worked with it before, I **highly recommend** reading [the article](https://blog.golang.org/laws-of-reflection) by the developers of go, giving the introduction to the library)~~
- HTTP server with the most basic pages and one API endpoint (I am planning to add some documentation of my API in a separate PR to make it really easy for front-end developers to integrate with the back-end)

### CODE REVIEW

File: config/postgres_credentials_example.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+user=USER password=PASSWORD dbname=DBNAME
```
Comment (DarkSquirrelComes) [inline config/postgres_credentials_example.txt:1]: Why you choose such config format?

File: src/database_bridge.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+func createConnection() (*sql.DB, error) {
+	return sql.Open("postgres", postgresCredentials)
+}
+
+// Sends `query` to the database and writes queried results to `result` (which should be <b>a pointer</b> to either a
+// slice (in this case the queried data is appended to its end) or a struct (then only the first selected value is
+// written to it, others are suppressed)). If `result` is a pointer to a struct, but no results are returned by the
+// query, the appropriate error is returned, `result` is not updated
+func dbQuery(result interface{}, query string, args ...interface{}) error {
```
Comment (DarkSquirrelComes) [inline src/database_bridge.go]: As i know, sql.Open(...) creates a connection pool and you shouldn't re-create it every time. Check it please.

File: src/common.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package main
+
+//Panics if `err` is not `nil`
+//
+// @param err Error to be checked for equality to `nil`
+func panicIfError(err error) {
```
Comment (DarkSquirrelComes) [inline src/common.go:8]: May be we should write some log here? And is it good error-handling pattern? (I really dont know. But i know a guy, who knows, i'll ask him)

File: src/webhandlers.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+		for idx, _ := range ProportionedTotalColor {
+			absoluteColor := ProportionedTotalColor[idx] * MaxProportion / TotalProportion
+			day.AverageColor[idx] = int(math.Round(absoluteColor))
+		}
+	}
+
+	js, err := json.Marshal(days)
+	panicIfError(err)
+	w.Header().Set("Content-Type", "application/json")
+	_, _ = w.Write(js)
```
Comment (DarkSquirrelComes) [inline src/webhandlers.go]: _, _

File: src/webhandlers.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+func handleLogout(w http.ResponseWriter, r *http.Request) {
+	session, _ := cookieStorage.Get(r, "session")
+	session.Options.MaxAge = -1
+	panicIfError(session.Save(r, w))
+
+	http.Redirect(w, r, "..", 302)
+}
+
+func handleApiDaysBrief(w http.ResponseWriter, r *http.Request) {
```
Comment (DarkSquirrelComes) [inline src/webhandlers.go]: Is size of this function ok?

**Review by DarkSquirrelComes — CHANGES_REQUESTED** [review]
- Efficient work with connection pool is the main issue here i see.
-  I'm not sure about config format
- Should we do something with error handling?
- All other small stuff in the comments)))

File: src/database_bridge.go (hunk 2 above)
Comment (kolayne) [reply src/database_bridge.go]: You are right, I'll fix it

File: src/common.go (hunk 3 above)
Comment (kolayne) [reply src/common.go:8]: I agree with logging. But I don't know about good patterns of error-handling patterns, it's just an idea a came up with, and this function was one of the reasons to request your review

File: src/webhandlers.go (hunk 4 above)
Comment (kolayne) [reply src/webhandlers.go]: Oops

File: src/webhandlers.go (hunk 5 above)
Comment (kolayne) [reply src/webhandlers.go]: I'll try to do something with it
UPD: should be better now

File: config/postgres_credentials_example.txt (hunk 1 above)
Comment (kolayne) [reply config/postgres_credentials_example.txt:1]: That's the format accepted by the [`PostgreSQL` driver](https://github.com/lib/pq) we are using. Is there a reason to use a different format and then convert?

File: src/webhandlers.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+var tmpl *template.Template
+var cookieStorage *sessions.CookieStore
+var requestsDecoder = schema.NewDecoder()
+var MaxProportion = 100.
+
+func loadHtmlTemplates() (*template.Template, error) {
+	if err := os.Chdir("src/templates"); err != nil {
+		return nil, err
+	}
+	defer os.Chdir("../..")
```
Comment (DarkSquirrelComes) [inline src/webhandlers.go]: WTF?

**Comment by kolayne** [comment]
Hi! I've made some fixes and decided to rewrite the history, so I had to force-push the branch and the history could change significantly (I say "could" because it did change significantly in my local version, but I don't remember which commits were pushed to the remote branch, so probably there is no big difference for you), sorry about that. Check out the improvements!

**Review by VladimirKoryakin — APPROVED** [review]
I think everything is fine except of probable differences in the database scheme and objects.go if the PR #12 will be merged earlier. But for the current branch master I approve these changes.

**Review by DarkSquirrelComes — APPROVED** [review]

**Review by gccenin — APPROVED** [review]

**Pull request merged**